walletnode/services/artworksearch/task.go

```diff
@@ -0,0 +1,89 @@
+package artworksearch
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pastelnetwork/gonode/walletnode/api/gen/artworks"
+
+	"github.com/pastelnetwork/gonode/common/log"
+	"github.com/pastelnetwork/gonode/common/service/task"
+	"github.com/pastelnetwork/gonode/pastel"
+)
+
+// Task is the task of searching for artwork.
+type Task struct {
+	task.Task
+	*Service
+
+	resultChan chan *pastel.Ticket
+	Request    *artworks.ArtSearchPayload
+}
+
+// Run starts the task
+func (task *Task) Run(ctx context.Context) error {
+	ctx = log.ContextWithPrefix(ctx, fmt.Sprintf("%s-%s", logPrefix, task.ID()))
+
+	log.WithContext(ctx).Debugf("Start task")
+	defer log.WithContext(ctx).Debugf("End task")
+
+	if err := task.run(ctx); err != nil {
+		log.WithContext(ctx).WithError(err).Warnf("Task failed")
+		return nil
+	}
+
+	return nil
+}
+
+func (task *Task) run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	_, err := task.actTickets(ctx)
+	if err != nil {
+		task.UpdateStatus(StatusTaskFailure)
+		return err
+	}
+
+	// TODO: iterate through tickets & get reg tickets
+	// In parallel; filter reg tickets based on search params
+	// & fetch thumbnail as soon as one matches
+	// send the thumbnail along with data through resultChan
+
+	return nil
+}
+
+// SubscribeSearchResult returns a new search resultof the state.
+func (task *Task) SubscribeSearchResult() chan *pastel.Ticket {
+	return task.resultChan
+}
+
+func (task *Task) actTickets(ctx context.Context) (tickets pastel.ActivationTickets, err error) {
+	actTickets, err := task.pastelClient.ActTickets(ctx, pastel.ActTicketAll, task.Request.MinBlock)
+	if err != nil {
+		return tickets, err
+	}
+
+	for i := 0; i < len(actTickets); i++ {
```

Comment:
range creates a copy of entire list which I try to avoid. range is readable but at the expense of time & memory
But I'll make this change

Reply:
I know, but it depends on the situation, actually I don't know how much data are you going to process.
But if we judge that you are going to receive them through the API, for sure that not hundreds of thousands, otherwise we should add a filter to the API itself